### PR TITLE
All pages

### PR DESCRIPTION
**Pull request by zFarrukh**

### CODE REVIEW

**Comment by vokanred** [comment]
Update the branch and resolve conflicts

**Comment by zFarrukh** [comment]
resolved

**Comment by vokanred** [comment]
https://user-images.githubusercontent.com/93642893/163759484-cd234e91-a2e5-454f-acec-36bf6334b3a3.mov

When I create the second account the form is initially red

**Comment by vokanred** [comment]
Screen Shot 2022-04-18 at 8 30 10 AM

NaN values in a table

**Comment by vokanred** [comment]
https://user-images.githubusercontent.com/93642893/163759786-f497534f-502b-456b-a945-4a1be14efa86.mov

Account details button is not shown sometimes

**Comment by vokanred** [comment]
Comments above just my observations that are not related to current PR. You can fix them when you want, but I think it's very important for demo.

Now starting to review the code

File: client/src/app/categories/categories-page/categories-page.component.ts [hunk 1]
```diff
@@ -1,15 +1,61 @@
...
+  }
+
+  openAddCategory() {
+    this.categoryService.addCategoryMode.next(true);
   }
 
+  onAddCategory(payload: { title: string; type: string }) {
+    this.categoryService.addCategory(payload).subscribe({
+      next: () => {
+        this.categoryService.getCategories().subscribe({
```
Comment (vokanred) [inline client/src/app/categories/categories-page/categories-page.component.ts]: Duplication

File: client/src/app/categories/categories-page/categories-page.component.ts [hunk 2]
```diff
@@ -1,15 +1,61 @@
...
+          next: (categories: ICategory[]) => {
+            this.categories = categories;
+          },
+        });
+      },
+    });
+  }
+
+  ngOnInit(): void {
+    this.categoryService.getCategories().subscribe({
```
Comment (vokanred) [inline client/src/app/categories/categories-page/categories-page.component.ts]: Duplication

Comment (vokanred) [reply client/src/app/categories/categories-page/categories-page.component.ts]: Create a method for getting categories

File: client/src/app/categories/categories-page/categories-page.component.ts [hunk 3]
```diff
@@ -1,15 +1,61 @@
...
+import { ICategory } from '../category.model';
+import { CategoryService } from '../category.service';
 
 @Component({
   selector: 'app-categories-page',
   templateUrl: './categories-page.component.html',
-  styleUrls: ['./categories-page.component.scss']
+  styleUrls: ['./categories-page.component.scss'],
 })
 export class CategoriesPageComponent implements OnInit {
```
Comment (vokanred) [inline client/src/app/categories/categories-page/categories-page.component.ts:13]: You should unsubscribe from every subscription

Comment (vokanred) [reply client/src/app/categories/categories-page/categories-page.component.ts:13]: You can use https://github.com/ngneat/until-destroy to make it easier

File: client/src/app/categories/categories-page/category-add/category-add.component.html [hunk 4]
```diff
@@ -0,0 +1,49 @@
+<mat-drawer-container
+  class="category"
+  autosize
+  [ngClass]="open ? '' : 'display-none'"
```
Comment (vokanred) [inline client/src/app/categories/categories-page/category-add/category-add.component.html]: [class.display-none]="!open"

Comment (vokanred) [reply client/src/app/categories/categories-page/category-add/category-add.component.html]: But anyway you shouldn't hide the drawer this way. The drawer has API for closing

File: client/src/app/categories/categories-page/category-add/category-add.component.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import { CategoryService } from '../../category.service';
+
+@Component({
+  selector: 'app-category-add',
+  templateUrl: './category-add.component.html',
+  styleUrls: ['./category-add.component.scss'],
+})
+export class CategoryAddComponent implements OnInit {
+  @Output() addCategory = new EventEmitter();
+  @ViewChild('drawer') drawer: any;
```
Comment (vokanred) [inline client/src/app/categories/categories-page/category-add/category-add.component.ts]: You don't need to call close() manually

Comment (vokanred) [reply client/src/app/categories/categories-page/category-add/category-add.component.ts]: Use [opened] input

File: client/src/app/categories/categories-page/category-add/category-add.component.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    this.addCategory.emit(this.categoryForm.value);
+    this.onClose();
+  }
+
+  constructor(private categoryService: CategoryService) {}
+
+  ngOnInit(): void {
+    this.categoryService.addCategoryMode.subscribe({
+      next: (category) => {
+        setTimeout(() => {
```
Comment (vokanred) [inline client/src/app/categories/categories-page/category-add/category-add.component.ts]: Don't call open() manually.

If you want to add setTimeout it's a good indicator that you're doing you are doing something wrong. :)

File: client/src/app/categories/categories-page/category-item/category-item.component.html [hunk 7]
```diff
@@ -0,0 +1,22 @@
+<div class="category" [ngClass]="category.type">
+  <form [formGroup]="categoryForm">
+    <input
+      class="category__title"
+      [value]="category.title"
```
Comment (vokanred) [inline client/src/app/categories/categories-page/category-item/category-item.component.html]: You shouldn't use [value] if you use formControl

File: client/src/app/categories/categories-page/category-item/category-item.component.html [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+      formControlName="title"
+      [ngClass]="isEditMode ? 'edit' : ''"
```
Comment (vokanred) [inline client/src/app/categories/categories-page/category-item/category-item.component.html]: [class.edit]="isEditMode"

File: client/src/app/categories/categories-page/category-item/category-item.component.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    const dialogRef = this.dialog.open(DialogComponent, {
+      width: '300px',
+      data: {
+        title: 'Category',
+        message: 'Are you sure you want to delete category?',
+        id: id,
+      },
+    });
+
+    dialogRef.afterClosed().subscribe((id) => {
```
Comment (vokanred) [inline client/src/app/categories/categories-page/category-item/category-item.component.ts]: unsubscribe

File: client/src/app/categories/category.service.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+import { Injectable } from '@angular/core';
+import { Observable, Subject, tap } from 'rxjs';
+import { environment } from 'src/environments/environment';
+import { ICategory } from './category.model';
+
+@Injectable({
+  providedIn: 'root',
+})
+export class CategoryService {
+  public categories: ICategory[] = [];
```
Comment (vokanred) [inline client/src/app/categories/category.service.ts]: You store categories here and also in categories page component

File: client/src/app/main/main-page/account/account-detail/account-detail.component.html [hunk 11]
```diff
@@ -0,0 +1,49 @@
+<mat-drawer-container
+  class="account"
+  autosize
+  [ngClass]="account ? '' : 'display-none'"
```
Comment (vokanred) [inline client/src/app/main/main-page/account/account-detail/account-detail.component.html]: the same

File: client/src/app/main/main-page/account/account-detail/account-detail.component.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+  templateUrl: './account-detail.component.html',
+  styleUrls: ['./account-detail.component.scss'],
+})
+export class AccountDetailComponent implements OnInit {
+  @Input() currency!: string;
+  @ViewChild('drawer') drawer: any;
+  account: IAccount | null = null;
+  selectedAccount: IAccount | null = null;
+  onClose(): void {
+    this.drawer.close();
```
Comment (vokanred) [inline client/src/app/main/main-page/account/account-detail/account-detail.component.ts]: same

File: client/src/app/main/main-page/account/account-detail/account-detail.component.ts [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+import { DialogComponent } from 'src/app/shared/dialog/dialog.component';
+import { IAccount } from '../account.model';
+import { AccountService } from '../account.service';
+
+@Component({
+  selector: 'app-account-detail',
+  templateUrl: './account-detail.component.html',
+  styleUrls: ['./account-detail.component.scss'],
+})
+export class AccountDetailComponent implements OnInit {
```
Comment (vokanred) [inline client/src/app/main/main-page/account/account-detail/account-detail.component.ts:16]: Remove drawer ref

File: client/src/app/main/main-page/account/account-edit/account-edit.component.ts [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+  styleUrls: ['./account-edit.component.scss'],
+})
+export class AccountEditComponent implements OnInit {
+  @Output() addAccount = new EventEmitter<IAccount>();
+  currencies = currency_list;
+  categories!: ICategory[] | null;
+  selectedAccount!: IAccount;
+  account: IAccount | null = null;
+  isEditMode = false;
+  @ViewChild('drawer') drawer: any;
```
Comment (vokanred) [inline client/src/app/main/main-page/account/account-edit/account-edit.component.ts]: Remove drawer ref

**Review by vokanred — CHANGES_REQUESTED** [review]
I haven't completed the review.

Fix all drawers. Use Drawer API instead of custom CSS.
Also you should unsubscribe everywhere but it is not so important for now.

**Comment by vokanred** [comment]
I see that you've implemented background for drawer. That is great but rawer container already has implementation for that. 

@Input()
hasBackdrop: boolean

**Comment by vokanred** [comment]
Every material component has a page with examples where you can find your case.

https://material.angular.io/components/sidenav/examples

**Review by vokanred — APPROVED** [review]
Approved

**Pull request merged**